src/main/java/com/aws/greengrass/logmanager/CloudWatchAttemptLogsProcessor.java

```diff
@@ -124,22 +125,31 @@ public CloudWatchAttempt processLogFiles(ComponentLogFileInformation componentLo
                 .replace("{componentName}", componentLogFileInformation.getName());
         attempt.setLogGroupName(logGroupName);
         String logStreamName = getLogStreamName(thingName);
-        LogFileGroup logFileGroup = componentLogFileInformation.getLogFileGroup();
         // Run the loop until all the log files from the component have been read or the max message
         // size has been reached.
         while (!componentLogFileInformation.getLogFileInformationList().isEmpty() && !reachedMaxSize.get()) {
             LogFileInformation logFileInformation = componentLogFileInformation.getLogFileInformationList().get(0);
             LogFile logFile = logFileInformation.getLogFile();
-            long startPosition = logFileInformation.getStartPosition();
             String fileHash = logFileInformation.getFileHash();
+            long startPosition = logFileInformation.getStartPosition();
+            LogFileGroup logFileGroup = componentLogFileInformation.getLogFileGroup();
+            // If the race condition happen, meaning file get rotated between scanning file and reading file, then
+            // let us update logFileGroup and get the new file object.
+            if (!fileHash.equals(logFile.hashString())) {
+                try {
+                    logFileGroup = logFileGroup.syncDirectory();
+                    logFile = logFileGroup.getFile(fileHash);
+                } catch (InvalidLogGroupException e) {
+                    logger.atDebug().cause(e).log("Invalid log group");
+                }
+            }
```

Comment:
THis is the fix for the NPE bug causing the LM breaks: 
WHy it happen: 
The file object is passed to this class after scanning and creating the logFileGroup. However, it got rotated before we create the bufferReader for it. Although it depends on how customers rotate it, but for current rotating policy, the file object now points to another file, that is the newly-created active file. THen we have the bug of the startPos mismatch. 

How it fix:
If the fileHash we pass from the scanning steps match the current file object, then it is not rotated. We keep the same. 
Otherwise, we will sync the logFIleGroup, and use the fileHash as the identifier to links to the target file.

Reply:
Note: 
This change still allow the possibility of the NPE happen, even though it is super low but not 0. 
Another addition fix for it is to handle the NPE nicely so we just simply skip this file if the startPos mismatch happen. I didn`t apply this fix so far is to intentionally test if the current fix reduce dramtically the possibility of the issue. Will apply the second fix once we confirm the first fix is correct.